crates/accelerate/src/qi/versor_u2.rs

```diff
@@ -0,0 +1,547 @@
+        Self(Unit::new_unchecked(Quaternion::new(cos, -sin, 0., 0.)))
+    }
+
+    /// Get the representation of an `RY(angle)` gate.
+    ///
+    /// Internally this calculates `sin` and `cos`, so discrete-angle forms will be more efficient
+    /// to be written explicitly.
+    #[inline]
+    fn from_ry(angle: f64) -> Self {
+        let (sin, cos) = (angle * 0.5).sin_cos();
+        Self(Unit::new_unchecked(Quaternion::new(cos, 0., -sin, 0.)))
+    }
+
+    /// Get the representation of an `RX(angle)` gate.
+    ///
+    /// Internally this calculates `sin` and `cos`, so discrete-angle forms will be more efficient
+    /// to be written explicitly.
+    #[inline]
+    fn from_rx(angle: f64) -> Self {
+        let (sin, cos) = (angle * 0.5).sin_cos();
+        Self(Unit::new_unchecked(Quaternion::new(cos, 0., 0., -sin)))
+    }
+
+    /// Fill a pre-allocated 2x2 Rust-native contiguous array with the Z-basis representation of
+    /// this versor.
+    #[inline]
+    pub fn matrix_contiguous_into(&self, matrix: &mut [[Complex64; 2]; 2]) {
+        let q = self.0.quaternion();
+        matrix[0][0] = Complex64::new(q.w, q.i);
+        matrix[0][1] = Complex64::new(q.j, q.k);
+        matrix[1][0] = Complex64::new(-q.j, q.k);
+        matrix[1][1] = Complex64::new(q.w, -q.i);
+    }
+}
+
+/// A versor-based (unit-quaternion) representation of a single-qubit gate.
+///
+/// In general, a single-qubit gate is a member of the group $U(2)$, and the group $SU(2)$ is
+/// isomoprhic to the versors.  We can keep track of the complex phase separately to the rotation
+/// action, to fully describe a member of $U(2)$.
+///
+/// See [VersorSU2] for the underlying quaternion representation.
+#[derive(Clone, Copy, Default, Debug, PartialEq)]
+pub struct VersorU2 {
+    /// The phase of the gate.  This should be multiplied as `exp(i * phase) * matrix(su2)` to
+    /// retrieve the explicit matrix form.
+    pub phase: f64,
+    /// The element of $SU(2)$ that has the same action as this gate, up to a global phase.
+    pub su2: VersorSU2,
+}
+impl VersorU2 {
+    /// Get the identity operator in quaternion form.
+    #[inline]
+    pub fn identity() -> Self {
+        Self {
+            phase: 0.,
+            su2: VersorSU2::identity(),
+        }
+    }
+
+    /// Get the versor representation of a 1q [StandardGate] without constructing a matrix.
+    ///
+    /// Returns the error state if `gate` is not 1q, or if any of the parameters are symbolic.
+    pub fn from_standard(gate: StandardGate, params: &[Param]) -> Result<Self, VersorU2Error> {
+        match gate {
+            StandardGate::GlobalPhaseGate => {
+                let &[Param::Float(phase)] = params else {
+                    return Err(VersorU2Error::Symbolic);
+                };
+                Ok(VersorSU2::identity().with_phase(phase))
+            }
+            StandardGate::HGate => {
+                debug_assert!(params.is_empty());
```

Comment:
We probably should get in the habit of doing this more for underlying assumptions like this.. Not that we build a ton in debug mode, but we do in some places so this would be good to sanity check.